joinrun/src/test/scala/code/chymyst/test/Patterns01Spec.scala

```diff
@@ -0,0 +1,101 @@
+package code.chymyst.test
+
+import java.util.concurrent.ConcurrentLinkedQueue
+
+import scala.collection.JavaConverters.asScalaIteratorConverter
+
+import code.chymyst.jc._
+import org.scalatest.{BeforeAndAfterEach, FlatSpec, Matchers}
+
+class Patterns01Spec extends FlatSpec with Matchers with BeforeAndAfterEach {
+
+  var tp: Pool = _
+
+  override def beforeEach(): Unit = {
+    tp = new SmartPool(4)
+  }
+
+  override def afterEach(): Unit = {
+    tp.shutdownNow()
+  }
+
+  behavior of "Chymyst"
+
+  it should "implement rendezvous" in {
+    val barrier1 = b[Unit,Unit]
+    val barrier2 = b[Unit,Unit]
+
+    val begin1 = m[Unit]
+    val begin2 = m[Unit]
+
+    val end1 = m[Unit]
+    val end2 = m[Unit]
+    val done = b[Unit, Unit]
+
+    val logFile = new ConcurrentLinkedQueue[String]
+
+    def f1() = logFile.add("f1")
+    def f2() = logFile.add("f2")
+    def g1() = logFile.add("g1")
+    def g2() = logFile.add("g2")
+
+    site(tp)(
+      go { case begin1(_) => f1(); barrier1(); g1(); end1() },
+      go { case begin2(_) => f2(); barrier2(); g2(); end2() },
+      go { case barrier1(_, r1) + barrier2(_, r2) => r1(); r2() },
+      go { case end1(_) + end2(_) + done(_, r) => r() }
+    )
+
+    begin1() + begin2()
+    done()
+    val result: Seq[String] = logFile.iterator().asScala.toSeq
+    // now, there must be f1 and f2 (in any order) before g1 and g2 (also in any order)
```

Comment:
possibly add for clarity hence we use Set below as opposed to Seq when comparing result